upstairs/src/lib.rs

```diff
@@ -4401,14 +4396,62 @@ impl Upstairs {
         let next_id = downstairs.next_id();
         let mut cur_offset: usize = 0;
 
-        let mut dep = downstairs.active.keys().cloned().collect::<Vec<u64>>();
-        dep.sort_unstable();
+        /*
+         * Walk the downstairs work active list, and pull out all the active
+         * jobs (anything we have not submitted back to the guest). Construct
+         * a list of dependencies for this write based on the
+         * following rules:
+         *
+         * - ignore everything that happened before the last flush
+         * - writes have to depend on the last flush completing
+         * - any overlap of impacted blocks requires a dependency
+         *
+         * TODO: scanning backwards, any overlap of impacted blocks will
+         * create a dependency. take this an example:
+         *
+         *       block
+         * op# | 0 1 2 | deps
+         * ----|-------------
+         *   0 | W     |
+         *   1 | W     | 0
+         *   2 | W     | 0,1
+         *
+         * op 2 depends on both op 1 and op 0. if dependencies are transitive
+         * with an existing job, it would be nice if those were removed from
+         * this job's dependencies.
+         */
+        let mut dep: Vec<u64> = Vec::with_capacity(200); // max in-flight jobs
+
+        // Search backwards in the list of active jobs
+        for job_id in downstairs
+            .active
+            .keys()
+            .sorted()
+            .collect::<Vec<&u64>>()
+            .iter()
+            .rev()
+        {
+            let job = &downstairs.active[job_id];
+
+            // Depend on the last flush, then break - flushes are a barrier for
+            // all writes.
+            if job.work.is_flush() {
+                dep.push(**job_id);
+                break;
+            }
+
+            // If this job impacts the same blocks as something already active,
+            // create a dependency.
+            if impacted_blocks.conflicts(&job.impacted_blocks) {
+                dep.push(**job_id);
+            }
+        }
 
         let mut writes: Vec<crucible_protocol::Write> =
-            Vec::with_capacity(nwo.len());
+            Vec::with_capacity(impacted_blocks.tuples().len());
 
         /* Lock here, through both jobs submitted */
```

Comment:
This comment should be removed, no longer applicable.

Reply:
done in 9a667b6